# Snakemake workflow: `map_metaT`

[![Snakemake](https://img.shields.io/badge/snakemake-â‰¥6.3.0-brightgreen.svg)](https://snakemake.github.io)
[![GitHub actions status](https://github.com/johnne/map_metaT/workflows/Tests/badge.svg?branch=main)](https://github.com/johnne/map_metaT/actions?query=branch%3Amain+workflow%3ATests)


A Snakemake workflow for mapping metatranscriptomic reads against a metagenomic reference.


## Usage

The usage of this workflow is described in the [Snakemake Workflow Catalog](https://snakemake.github.io/snakemake-workflow-catalog/?usage=johnne%2Fmap_metaT).

If you use this workflow in a paper, don't forget to give credits to the authors by citing the URL of this (original) repository and its DOI (see above).

# TODO

* Replace `<owner>` and `<repo>` everywhere in the template (also under .github/workflows) with the correct `<repo>` name and owning user or organization.
* Replace `<name>` with the workflow name (can be the same as `<repo>`).
